File: Java/Pilha.java
public class Pilha extends Lista {
    private Node inicio;
    public Pilha(){
        this.inicio = null;
    }

    @Override
    public void insere(int value) {
        // Inserir no topo da pilha (push)
        Node newNode = new Node(value);
        newNode.setNext(this.inicio);
        this.inicio = newNode;
        this.setSize(this.getSize() + 1);
    }
    
    @Override
    public void retira() {
        // Retirar do topo da pilha (pop)
        if (this.inicio != null) {
            this.inicio = this.inicio.getNext();
            this.setSize(this.getSize() - 1);
        }
    }

    @Override
    public int primeiro() {
        // Retorna o topo da pilha
        if (this.getSize() == 0) {
            return -1;
        }
        return this.inicio.getValue();
    }

    @Override
    public int ultimo() {
        // Não é relevante para uma pilha, pode retornar -1 ou lançar uma exceção
        return -1;
    }

    @Override
    public void retira(int value) {
        throw new UnsupportedOperationException("Unimplemented method 'retira'");
    }
}

File: C++/Pilha.cpp
#include "Pilha.h"

Pilha::Pilha() : Lista(), top(nullptr) {}


void Pilha::insere(int value) {
    // Adiciona um elemento no topo da pilha
    Node* novo = new Node(value);
    novo->next = top;
    top = novo;
}

void Pilha::retira(int value) {
    if (top == nullptr) {
        // Pilha vazia
        return;
    }
        Node* temp = top;
        top = top->next;
        delete temp;
}

int Pilha::primeiro() {
    if (top == nullptr) {
        return -1; // Ou algum valor indicando erro/vazio
    }

    return top->value;
}

int Pilha::ultimo() {
    if (top == nullptr) {
        // Pilha vazia
        return -1;
    }

    Node* temp = top;
    while (temp->next != nullptr) {
        temp = temp->next;
    }

    return temp->value;
}